src/java/org/apache/cassandra/gms/Gossiper.java

```diff
@@ -1506,8 +1496,20 @@ private void applyNewStates(InetAddressAndPort addr, EndpointState localState, E
         }
         localState.addApplicationStates(updatedStates);
 
+        // get rid of legacy fields once the cluster is not in mixed mode
+        if (!hasMajorVersion3Nodes())
+            localState.removeMajorVersion3LegacyApplicationStates();
+
         for (Entry<ApplicationState, VersionedValue> updatedEntry : updatedStates)
+        {
+            // filters out legacy change notifications
+            // only if local state already indicates that the peer has the new fields
+            if ((INTERNAL_IP == updatedEntry.getKey() && localState.containsApplicationState(INTERNAL_ADDRESS_AND_PORT))
+                ||(STATUS == updatedEntry.getKey() && localState.containsApplicationState(STATUS_WITH_PORT))
+                || (RPC_ADDRESS == updatedEntry.getKey() && localState.containsApplicationState(NATIVE_ADDRESS_AND_PORT)))
+                continue;
```

Comment:
i think this needs to be checks against the remoteState
```suggestion
            // filters out duplicated legacy change notifications to avoid deadlock of the OutboundConnection event loop thread
            if (hasMajorVersion3Nodes() && (INTERNAL_IP == updatedEntry.getKey() && remoteState.containsApplicationState(INTERNAL_ADDRESS_AND_PORT))
                ||(STATUS == updatedEntry.getKey() && remoteState.containsApplicationState(STATUS_WITH_PORT))
                || (RPC_ADDRESS == updatedEntry.getKey() && remoteState.containsApplicationState(NATIVE_ADDRESS_AND_PORT)))
                continue;
```

"deadlock of the OutboundConnection event loop thread" is described [here](https://issues.apache.org/jira/browse/CASSANDRA-16381?focusedCommentId=17288680&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17288680).

Reply:
I do not think the suggested change is necessary. 

At the point of firing notifications. 
The `remoteState` has been merged with `localState`. 
The legacy fields has been removed if possible. 
So checking against the new fields in the `localState` has the same effect but it no longer needs to check if it is in mixed mode again. 

I have also added `org.apache.cassandra.gms.GossiperTest#testNotFireDuplicatedNotificationsWithUpdateContainsOldAndNewState` to verify the behavior.

Reply:
you are absolutely right :)